### ISSUE TITLE
During ENOSPC nemesis, scylla coredumped

### ISSUE THREAD

**Issue by fruch** [scylladb/scylla#4614]
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [X] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash):  3.1.0.rc2-0.20190624.433cb93f7.el7.x86_64
Cluster size: 4
OS (RHEL/CentOS/Ubuntu/AWS AMI): ami-0ac07c9630d9e4eea

*Hardware details (for performance issues)*          Delete if unneeded
Platform (physical/VM/cloud instance type/docker): AWS

*test_id*: fed493bf-339e-41ab-ab78-f710fb322385
*node log*:  http://scratch.scylladb.com/fruch/scylla_issue_4614_node.log

```
          PID: 6688 (scylla)
           UID: 996 (scylla)
           GID: 1001 (scylla)
        Signal: 6 (ABRT)
     Timestamp: Thu 2019-06-27 05:36:49 UTC (1h 5min ago)
  Command Line: /usr/bin/scylla /opt/scylladb/bin/../libexec/scylla.bin --blocked-reactor-notify-ms 500 --abort-on-lsa-bad-alloc 1 --abort-on-seastar-bad-alloc --log-to-syslog 1 --log-to-stdout 0 --default-log-level info --network-stack posix --io-properties-file=/etc/scylla.d/io_properties.yaml --cpuset 1-7,9-15
    Executable: /opt/scylladb/libreloc/ld.so
 Control Group: /system.slice/scylla-server.service
          Unit: scylla-server.service
         Slice: system.slice
       Boot ID: aa6c40088f3d4b7b82f3519aafb751f7
    Machine ID: df877a200226bc47d06f26dae0736ec9
      Hostname: ip-10-0-89-228.eu-west-1.compute.internal
       Message: Process 6688 (scylla) of user 996 dumped core.
```

```
void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at /usr/include/boost/program_options/variables_ma
p.hpp:146                                                                                                                                                                                                                         
seastar::backtrace_buffer::append_backtrace() at /usr/include/boost/program_options/variables_map.hpp:146                                                                                                                         
 (inlined by) print_with_backtrace at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/build/release/../../src/core/reactor.cc:1104                                                                             
seastar::print_with_backtrace(char const*) at /usr/include/boost/program_options/variables_map.hpp:146                                                                                                                            
sigabrt_action at /usr/include/boost/program_options/variables_map.hpp:146                                                                                                                                                        
 (inlined by) operator() at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/build/release/../../src/core/reactor.cc:5007                                                                                       
 (inlined by) _FUN at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/build/release/../../src/core/reactor.cc:5003                                                                                             
?? ??:0                                                                                                                                                                                                                           
?? ??:0                                                                                                                                                                                                                           
?? ??:0                                                                                                                                                                                                                           
?? ??:0                                                                                                                                                                                                                           
?? ??:0                                                                                                                                                                                                                           
void seastar::future_state<>::set<>() at /usr/include/boost/program_options/variables_map.hpp:146                                                                                                                                 
void seastar::future_state<>::set<>() at /usr/include/boost/program_options/variables_map.hpp:146                                                                                                                                 
 (inlined by) void seastar::promise<>::set_value<>() at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/build/release/../../include/seastar/core/future.hh:453                                                 
 (inlined by) seastar::append_challenged_posix_file_impl::process_queue() at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/build/release/../../src/core/reactor.cc:2540                                      
seastar::append_challenged_posix_file_impl::close() at /usr/include/boost/program_options/variables_map.hpp:146                                                                                                                   
auto do_io_check<checked_file_impl::close()::{lambda()#1}, , void>(std::function<void (std::__exception_ptr::exception_ptr)> const&, checked_file_impl::close()::{lambda()#1}&&) at crtstuff.c:?                                  
 (inlined by) auto do_io_check<checked_file_impl::close()::{lambda()#1}, , void>(std::function<void (std::__exception_ptr::exception_ptr)> const&, checked_file_impl::close()::{lambda()#1}&&) at /jenkins/slave/workspace/scylla-
3.1/relocatable-pkg/scylla/disk-error-handler.hh:73                                                                                                                                                                               
checked_file_impl::close() at crtstuff.c:?                                                                                                                                                                                        
seastar::file::close() at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/build/release/../../include/seastar/core/file.hh:395                                                                                 
 (inlined by) seastar::file_data_sink_impl::close()::{lambda()#1}::operator()() const at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/build/release/../../src/core/fstream.cc:428                           
 (inlined by) seastar::future<> seastar::futurize<seastar::future<> >::apply<seastar::file_data_sink_impl::close()::{lambda()#1}&>(seastar::file_data_sink_impl::close()::{lambda()#1}&) at /jenkins/slave/workspace/scylla-3.1/re
locatable-pkg/scylla/seastar/build/release/../../include/seastar/core/future.hh:1402                                                                                                                                              
 (inlined by) seastar::future<>::finally_body<seastar::file_data_sink_impl::close()::{lambda()#1}, true>::operator()(seastar::future<>&&) at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/build/release/../.
./include/seastar/core/future.hh:1078                                                                                                                                                                                             
seastar::future<> seastar::futurize<seastar::future<> >::apply<seastar::future<>::finally_body<seastar::file_data_sink_impl::close()::{lambda()#1}, true>, seastar::future<> >(seastar::future<>::finally_body<seastar::file_data_
sink_impl::close()::{lambda()#1}, true>&&, seastar::future<>&&) at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/build/release/../../include/seastar/core/future.hh:1402                                     
 (inlined by) seastar::future<> seastar::future<>::then_wrapped_impl<seastar::future<>::finally_body<seastar::file_data_sink_impl::close()::{lambda()#1}, true>, seastar::future<> >(seastar::future<>::finally_body<seastar::file
_data_sink_impl::close()::{lambda()#1}, true>&&) at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/build/release/../../include/seastar/core/future.hh:994                                                     
 (inlined by) seastar::future<> seastar::future<>::then_wrapped<seastar::future<>::finally_body<seastar::file_data_sink_impl::close()::{lambda()#1}, true>, seastar::future<> >(seastar::future<>::finally_body<seastar::file_data
_sink_impl::close()::{lambda()#1}, true>&&) at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/build/release/../../include/seastar/core/future.hh:978                                                          
 (inlined by) seastar::future<> seastar::future<>::finally<seastar::file_data_sink_impl::close()::{lambda()#1}>(seastar::file_data_sink_impl::close()::{lambda()#1}&&) at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scyl
la/seastar/build/release/../../include/seastar/core/future.hh:1051                                                                                                                                                                
 (inlined by) seastar::file_data_sink_impl::close() at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/build/release/../../src/core/fstream.cc:429                                                             
seastar::data_sink::close() at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/include/seastar/core/sstring.hh:257
 (inlined by) seastar::output_stream<char>::close()::{lambda()#3}::operator()() const at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/include/seastar/core/iostream-impl.hh:488
 (inlined by) seastar::future<> seastar::futurize<seastar::future<> >::apply<seastar::output_stream<char>::close()::{lambda()#3}&>(seastar::output_stream<char>::close()::{lambda()#3}&) at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/include/seastar/core/future.hh:1402
 (inlined by) seastar::future<>::finally_body<seastar::output_stream<char>::close()::{lambda()#3}, true>::operator()(seastar::future<>&&) at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/include/seastar/core/future.hh:1078
seastar::future<> seastar::futurize<seastar::future<> >::apply<seastar::future<>::finally_body<seastar::output_stream<char>::close()::{lambda()#3}, true>, seastar::future<> >(seastar::future<>::finally_body<seastar::output_stream<char>::close()::{lambda()#3}, true>&&, seastar::future<>&&) at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/include/seastar/core/sstring.hh:257
 (inlined by) seastar::future<> seastar::future<>::then_wrapped_impl<seastar::future<>::finally_body<seastar::output_stream<char>::close()::{lambda()#3}, true>, seastar::future<> >(seastar::future<>::finally_body<seastar::output_stream<char>::close()::{lambda()#3}, true>&&) at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/include/seastar/core/future.hh:994
 (inlined by) seastar::future<> seastar::future<>::then_wrapped<seastar::future<>::finally_body<seastar::output_stream<char>::close()::{lambda()#3}, true>, seastar::future<> >(seastar::future<>::finally_body<seastar::output_stream<char>::close()::{lambda()#3}, true>&&) at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/include/seastar/core/future.hh:978
 (inlined by) seastar::future<> seastar::future<>::finally<seastar::output_stream<char>::close()::{lambda()#3}>(seastar::output_stream<char>::close()::{lambda()#3}&&) at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/include/seastar/core/future.hh:1051
 (inlined by) seastar::output_stream<char>::close() at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/include/seastar/core/iostream-impl.hh:489
sstables::file_writer::close() at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/include/seastar/core/future.hh:768
 (inlined by) operator()<std::optional<sstables::file_writer> > at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/sstables/mc/writer.cc:773
 (inlined by) sstables::mc::writer::~writer() at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/sstables/mc/writer.cc:779
sstables::mc::writer::~writer() at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/include/seastar/core/future.hh:768
flat_mutation_reader::impl::consumer_adapter<sstables::sstable_writer>::~consumer_adapter() at crtstuff.c:?
 (inlined by) std::unique_ptr<sstables::sstable_writer::writer_impl, std::default_delete<sstables::sstable_writer::writer_impl> >::~unique_ptr() at /usr/include/c++/8/bits/unique_ptr.h:274
 (inlined by) sstables::sstable_writer::~sstable_writer() at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/./sstables/sstables.hh:895
 (inlined by) flat_mutation_reader::impl::consumer_adapter<sstables::sstable_writer>::~consumer_adapter() at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/./flat_mutation_reader.hh:211
seastar::noncopyable_function<void ()>::direct_vtable_for<seastar::futurize<std::result_of<std::decay<sstables::sstable::write_components(flat_mutation_reader, unsigned long, seastar::lw_shared_ptr<schema const>, sstables::sstable_writer_config const&, encoding_stats, seastar::io_priority_class const&)::{lambda()#1}>::type ()>::type>::type seastar::async<sstables::sstable::write_components(flat_mutation_reader, unsigned long, seastar::lw_shared_ptr<schema const>, sstables::sstable_writer_config const&, encoding_stats, seastar::io_priority_class const&)::{lambda()#1}>(seastar::thread_attributes, std::decay&&, (std::decay<sstables::sstable::write_components(flat_mutation_reader, unsigned long, seastar::lw_shared_ptr<schema const>, sstables::sstable_writer_config const&, encoding_stats, seastar::io_priority_class const&)::{lambda()#1}>::type&&)...)::{lambda(seastar::futurize<std::result_of<std::decay<{lambda()#1}>::type ()>::type> seastar::async<{lambda()#1}>(seastar::futurize<std::result_of<std::decay<{lambda()#1}>::type ()>::type>::type, seastar::thread_attributes, std::decay<{lambda()#1}>::type&&)::work&)#1}::operator()(seastar::futurize<std::result_of<std::decay<{lambda()#1}>::type ()>::type> seastar::async<{lambda()#1}>(seastar::futurize<std::result_of<std::decay<{lambda()#1}>::type ()>::type>::type, seastar::thread_attributes, std::decay<{lambda()#1}>::type&&)::work)::{lambda()#1}>::call(seastar::noncopyable_function<void ()> const*) [clone .cold.8356] at sstables.cc:?
 (inlined by) auto flat_mutation_reader::consume_in_thread<sstables::sstable_writer, auto flat_mutation_reader::consume_in_thread<sstables::sstable_writer>(sstables::sstable_writer, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(dht::decorated_key const&)#1}>(sstables::sstable_writer, auto flat_mutation_reader::consume_in_thread<sstables::sstable_writer>(sstables::sstable_writer, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(dht::decorated_key const&)#1}, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >) at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/./flat_mutation_reader.hh:368
 (inlined by) auto flat_mutation_reader::consume_in_thread<sstables::sstable_writer>(sstables::sstable_writer, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >) at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/./flat_mutation_reader.hh:376
 (inlined by) operator() at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/sstables/sstables.cc:2427
 (inlined by) apply at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/include/seastar/core/apply.hh:35
 (inlined by) apply<sstables::sstable::write_components(flat_mutation_reader, uint64_t, schema_ptr, const sstables::sstable_writer_config&, encoding_stats, const seastar::io_priority_class&)::<lambda()> > at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/include/seastar/core/apply.hh:43
 (inlined by) apply_tuple<sstables::sstable::write_components(flat_mutation_reader, uint64_t, schema_ptr, const sstables::sstable_writer_config&, encoding_stats, const seastar::io_priority_class&)::<lambda()> > at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/include/seastar/core/future.hh:1346
 (inlined by) apply<sstables::sstable::write_components(flat_mutation_reader, uint64_t, schema_ptr, const sstables::sstable_writer_config&, encoding_stats, const seastar::io_priority_class&)::<lambda()> > at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/include/seastar/core/future.hh:1380
 (inlined by) operator() at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/include/seastar/core/thread.hh:324
 (inlined by) call at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/include/seastar/util/noncopyable_function.hh:71
seastar::noncopyable_function<void ()>::operator()() const at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/build/release/../../include/seastar/util/noncopyable_function.hh:145
 (inlined by) seastar::thread_context::main() at /jenkins/slave/workspace/scylla-3.1/relocatable-pkg/scylla/seastar/build/release/../../src/core/thread.cc:317
```

coredump wasn't saved cause of not enough disk space :
```
systemd-coredump[19365]: Not enough disk space for coredump of 6688 (scylla), refusing.
```

**Comment by fruch** [comment]
@slivne I remember seeing an issue somewhere (couldn't found it again) you've open about making the root disk of our AMIs bigger by default, is it related to why this coredump didn't had enough disk space ?

**Comment by bhalevy** [comment]
> @slivne I remember seeing an issue somewhere (couldn't found it again) you've open about making the root disk of our AMIs bigger by default, is it related to why this coredump didn't had enough disk space ?

You mentioned ENOSPC nemesis, could it be related? :-)

**Comment by fruch** [comment]
It's not the same disk, the root disk is the one that doesn't have enough space
```
Filesystem      1K-blocks      Used  Available Use% Mounted on
devtmpfs         62879816         0   62879816   0% /dev
tmpfs            62907004         0   62907004   0% /dev/shm
tmpfs            62907004     41488   62865516   1% /run
tmpfs            62907004         0   62907004   0% /sys/fs/cgroup
/dev/xvda1       15717356   7074132    8643224  46% /
tmpfs            12581404         0   12581404   0% /run/user/1000
/dev/md0       3708861024 448263504 3260597520  13% /var/lib/scylla
```

**Comment by slivne** [comment]
We change the system and have it use /var/lib/scylla for coredumps

So it is the same disk

**Comment by fruch** [comment]
oops... then it might be a problem to catch those during this specific nemesis.
we'll need to think about how to handle those.

@espindola this machine is still up, if you need it's details. ping me.

**Comment by slivne** [comment]
@bhalevy ping

**Comment by bhalevy** [comment]
@espindola can you please look into the stacktrace above and see if you can come up with a theory and/or a unit test reproducer?

**Comment by espindola** [comment]
Do we have the logs? I assume the abort was from

assert(_u.st == state::future);

in future_state::set, correct? A core file would be very handy. Is this easy to reproduce? If so, could we configure coredumps to go to another filesystem?

In any case, if the above assumption is correct, the only set_value that matches the backtrace is

        _completed.set_value();

In seastar::append_challenged_posix_file_impl::process_queue. This also seems to be the only location where it is set. And it can only be set if _done is true. Since we clear _done, I think this can only happen if close is called twice.

Maybe a way to test this is to change append_challenged_posix_file_impl::close to have

auto fut = _completed.get_future();
assert(!fut.available());
process_queue();
return fut.then(...)

And maybe a log saying that we are closing a given file.

@gleb-cloudius, you fixed the last bug around this area. Does the above sound reasonable?

**Comment by avikivity** [comment]
double-close is a plausible scenario for this

**Comment by espindola** [comment]
Looking up the stack, the first place I see where a double close could happen is if two files point to the same file_impl. That seems unlikely, but why does the file have a shared_ptr to file_impl if not for that?

Further up, in the mc writer, we have in  writer::consume_end_of_stream

  _index_writer->close();
  _index_writer.reset();
...
  close_data_writer(); // moves before close

These variables are std::optional<file_writer> or std::unique_ptr<flie_writer>. In the writer destructor they are closed without moving or reset.

If the destructor is running in parallel with another method, that would be a bug on its own, but it could manifest as a double close.

Maybe it would be better to organize the mc writer like the other stream/writer interfaces? That is:

* add a writer::close, that must be called by the user.
* in writer::close, we close and clear _index_writer and _data_writer
* in ~writer we assert that both are cleared

**Comment by fruch** [comment]
it's not easily reproduced, I.e. we seen it only once.

as for the coredump, it would be nice if we have a mount for the coredumps which isn't overrun by scylla data.

I'll upload the logs of that run.

**Comment by fruch** [comment]
@espindola 
here's the node log:
http://scratch.scylladb.com/fruch/scylla_issue_4614_node.log

**Comment by fruch** [comment]
@espindola
happened again: 
http://scratch.scylladb.com/fruch/scylla_issue_4614_node_2.log

we still didn't had the change to fix the coredump failing to write, so no much more new info. just a new log for reference

**Comment by slivne** [comment]
@espindola if you have a patch QA can run a test with your RPMS to check if this fixes the issues.

**Comment by espindola** [comment]
I posted "[PATCH] mc writer: More uniform closing of the file_writers"
which might avoid the double close, but there would still be something
odd going on.

Cheers,
Rafael

**Comment by fruch** [comment]
@slivne @roydahan, do we commit that just came in  said  

> This fixes a possible cause of #4614.

do we really want to close this issue ?

**Comment by bhalevy** [comment]
@fruch we need to verify the fix in any case so this issue can be reopened and hopefully re-closed after verification or reopened in case verification shows the issue is still not solved.
